File: typescript/src/1_3_8_Observer.ts
// WebSocketを使ったトランザクションステータスの監視を行うコード
import { PrivateKey } from "symbol-sdk"
import {
  Network,
  SymbolFacade,
  descriptors,
} from "symbol-sdk/symbol"

import dotenv from "dotenv"
import { createAndSendTransaction } from "../functions/createAndSendTransaction"

dotenv.config()

const NODE_URL = "https://sym-test-03.opening-line.jp:3001"
const facade = new SymbolFacade(Network.TESTNET)
const privateKeyA = new PrivateKey(process.env.PRIVATE_KEY_A!)
const accountA = facade.createAccount(privateKeyA)
const privateKeyB = new PrivateKey(process.env.PRIVATE_KEY_B!)
const accountB = facade.createAccount(privateKeyB)

// WebSocketクライアントの生成
const wsEndpoint = NODE_URL.replace("http", "ws") + "/ws"
let uid = ""
const subscriptions = new Map<string, Function[]>()

// サブスクリプション管理
const subscribe = (channel: string, callback: Function): void => {
  if (!subscriptions.has(channel)) {
    subscriptions.set(channel, [])
  }
  subscriptions.get(channel)?.push(callback)
}

// チャンネル設定
const confirmedChannelName = `confirmedAdded/${accountA.address}`
const unconfirmedChannelName = `unconfirmedAdded/${accountA.address}`

subscribe(confirmedChannelName, (tx: any) => {
  //承認済みトランザクションを検知した時の処理
  console.log("承認済みトランザクション:", tx)
  console.log("結果 Success", "エクスプローラー ", `https://testnet.symbol.fyi/transactions/${tx.meta.hash}`)
})
subscribe(unconfirmedChannelName, (tx: any) =>{
  //承認済みトランザクションを検知した時の処理
  console.log("未承認済みトランザクション:", tx)
})
// WebSocket接続の管理
const initializeWebSocket = () => {
  const ws = new WebSocket(wsEndpoint)

  // WebSocketに接続した時の処理
  ws.addEventListener("open", () => {
    console.log("WebSocket接続確立")
  })

  // WebSocketでメッセージを検知した時の処理
  ws.addEventListener("message", (event: MessageEvent) => {
    try {
      const message = JSON.parse(event.data)

      // UIDの初期処理
      if (message.uid) {
        uid = message.uid
        console.log("接続ID:", uid)
        // チャンネル購読
        ws.send(JSON.stringify({ uid, subscribe: confirmedChannelName }));
        ws.send(JSON.stringify({ uid, subscribe: unconfirmedChannelName }));
        return
      }

      // 購読チャンネルからのメッセージ処理
      const handlers = subscriptions.get(message.topic)
      handlers?.forEach((handler) => {
        handler(message.data)
        // confirmedAddedを検知したらWebSocketを切る処理
        if (message.topic === confirmedChannelName) {
          ws.close()
        }
      })
    } catch (err) {
      console.error("メッセージ処理エラー:", err)
    }
  })

  // WebSocketでエラーを検知した時の処理
  ws.addEventListener("error", (event) => {
    console.error("WebSocketエラー:", event)
  })

  // WebSocketが閉じた時の処理
  ws.addEventListener("close", () => {
    console.log("WebSocket接続終了")
    uid = ""
    subscriptions.clear()
  })

  return ws
}

// WebSocket開始
initializeWebSocket()

// 接続が確立するまで1秒待つ
await new Promise((resolve) => setTimeout(resolve, 1000))

// 監視で検知させるための転送トランザクション
const transferDescriptor =
  new descriptors.TransferTransactionV1Descriptor(
    accountB.address,
    [],
    "\0Hello, accountB!",
  )

await createAndSendTransaction(
  transferDescriptor,
  accountA
)

File: python/src/1_3_8_observer.py
import os
import sys
import json
import asyncio
import requests
from websockets.legacy.client import connect
from dotenv import load_dotenv
from symbolchain.CryptoTypes import PrivateKey
from symbolchain.facade.SymbolFacade import SymbolFacade
from symbolchain.sc import TransferTransactionV1
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from functions.send_transaction import send_transaction

async def initialize_websocket(NODE_URL, account_a) -> None:
    ws_endpoint = NODE_URL.replace("http", "ws") + "/ws"

    # WebSocket接続
    async with connect(ws_endpoint) as websocket:
        # 接続時のレスポンスからUIDを取得
        response_json = json.loads(await websocket.recv())
        uid = response_json['uid']
        print(f'接続ID: {uid}')

        # 監視チャンネルの設定
        confirmed_channel_name = f'confirmedAdded/{account_a.address}'
        unconfirmed_channel_name = f'unconfirmedAdded/{account_a.address}'

        await websocket.send(json.dumps({'uid': uid, 'subscribe': f'{confirmed_channel_name}/{account_a.address}'}))
        await websocket.send(json.dumps({'uid': uid, 'subscribe': f'{unconfirmed_channel_name}/{account_a.address}'}))

        # トランザクションの状態変化を監視
        try:
            while True:
                response_json = json.loads(await websocket.recv())
                topic = response_json['topic']
                tx = response_json['data']
                # 承認済みトランザクションを検知した時の処理
                if topic.startswith('confirmedAdded'):
                    print(f'承認トランザクション検知: {tx}')
                    print("結果 Success", "エクスプローラー ", f"https://testnet.symbol.fyi/transactions/{tx['meta']['hash']}")
                    break
                # 未承認済みトランザクションを検知した時の処理
                elif topic.startswith('unconfirmedAdded'):
                    print(f'未承認トランザクション検知: {tx}')
        except Exception as e:
            print("WebSocketエラー:", e)

        print("WebSocket接続終了")
        uid = ""
        await websocket.close()
            
async def main() -> None:
    load_dotenv()

    NODE_URL = "https://sym-test-03.opening-line.jp:3001"    
    facade = SymbolFacade('testnet')
    private_key_a = os.getenv('PRIVATE_KEY_A') or ""
    account_a = facade.create_account(PrivateKey(private_key_a))
    private_key_b = os.getenv('PRIVATE_KEY_B') or ""
    account_b = facade.create_account(PrivateKey(private_key_b))

    network_time = requests.get(f"{NODE_URL}/node/time").json()
    current_timestamp: int = int(network_time['communicationTimestamps']['receiveTimestamp'])
    deadline_timestamp: int = current_timestamp + (2 * 60 * 60 * 1000)
    
    # 監視対象のトランザクション送信
    transfer_tx: TransferTransactionV1 = facade.transaction_factory.create({
        'type': 'transfer_transaction_v1',
        'recipient_address': account_b.address,
        'mosaics': [],
        'message': b'\0Hello, accountB!',
        'signer_public_key': account_a.public_key,
        'deadline': deadline_timestamp
    })

    send_transaction(transfer_tx, account_a)

    await initialize_websocket(NODE_URL, account_a)

if __name__ == "__main__":
    asyncio.run(main())